### DISCUSSION TITLE
Should the result of average_precision_score in sklearn be the same as mAP in YOLO results?

### DISCUSSION THREAD

**Discussion by M-Amrollahi** [ultralytics/ultralytics#20386] [Q&A]
I have developed a module to draw the graph of precision-recall. The module gets the detections and ground truths ( both in yolo format) and then try to draw a precision-recall graph based on:
```python
sklearn.metrics: precision_recall_curve, average_precision_score
```
But when I compare the results with the mAP in YOLO results, it is kind of different. In both of the methods I use same IoU and conf threshold.
May I know first, if I can compare these two methods and their results and if so, why they are different? 

What I want at the end, is to draw the graph of PR based on the detections and ground truth labels.
Thanks

**Comment by UltralyticsAssistant** [comment]
👋 Hello @M-Amrollahi, thank you for your interest in Ultralytics 🚀! We recommend checking out the [Docs](https://docs.ultralytics.com/) where you can explore many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples—many common questions are already answered there.

This is an automated response to help you get started. An Ultralytics engineer will review your question and assist you soon!

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

For custom training or evaluation ❓ questions, please share as much detail as possible, including sample detection outputs, ground truth formats, code snippets, and any relevant logs. Also, take a look at our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) to ensure your pipeline is optimal.

Join the Ultralytics community where it suits you best! For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Reply by glenn-jocher** [reply]
Hi @M-Amrollahi,

The difference you're observing between sklearn's `average_precision_score` and YOLO's mAP calculation is expected. These two methods use different calculation approaches:

1. **Implementation differences**: Ultralytics implements a specific `compute_ap()` function for object detection that follows standard object detection protocols (like COCO or Pascal VOC), while sklearn's implementation is designed for general classification tasks.

2. **Interpolation methods**: YOLO uses either interpolated precision (default) or the continuous version to calculate the area under the precision-recall curve.

3. **Class handling**: YOLO's mAP properly averages across all classes with specific handling of object detection nuances.

If you want to match YOLO's exact PR curve calculation, I recommend using the Ultralytics implementation directly. You can find the relevant code in the `ultralytics/utils/metrics.py` file, specifically the `compute_ap()` and `ap_per_class()` functions. These functions handle all the object detection-specific calculations and produce the curves you see in YOLO's results.

For custom visualization, you could adapt those functions to work with your detection outputs while maintaining YOLO's calculation methodology.

**Reply by M-Amrollahi** [reply]
I just wanted to bring the numbers or any graph in my reports and I am just looking for any standard method that I can draw the PR graph for both my YOLO and COCO(json) model findings. That is why I asked it.
Thanks

**Reply by M-Amrollahi** [reply]
Thanks @glenn-jocher 
Actually, I could not use ```ap_per_class```. Could you tell me how can i use this function?
I have my annotations and my my detection txt files, and I want to use this function to get the mAP. I am not sure if we can have the precision or recall as well.
Thanks for the helps

**Reply by glenn-jocher** [reply]
@M-Amrollahi, to use the `ap_per_class()` function with your txt files, you'll need to:

1. First convert your annotations and detections into the format expected by `ap_per_class()`:
   - `tp`: Array of true positives
   - `conf`: Array of confidence scores
   - `pred_cls`: Array of predicted classes
   - `target_cls`: Array of target classes

Here's a simplified example of how you can use it:

```python
from ultralytics.utils.metrics import ap_per_class
import numpy as np

# Process your txt files to get these arrays
# Example processing (you'll need to adapt this for your specific format)
tp = np.array([...])  # Binary array where 1=true positive, 0=false positive
conf = np.array([...])  # Confidence scores from your detections
pred_cls = np.array([...])  # Class indices from your detections
target_cls = np.array([...])  # Class indices from your ground truth

# Call ap_per_class
results = ap_per_class(
    tp, 
    conf, 
    pred_cls, 
    target_cls,
    plot=True,  # Set to True to generate PR curves
    save_dir="path/to/save",  # Where to save the curve plots
    names={}  # Dictionary mapping class indices to names
)

# Extract metrics from results
precision, recall, ap, f1, ap_class = results[:5]
p_curve, r_curve, f1_curve = results[7:10]

# mAP50 is ap at index 0, mAP50-95 is the mean of ap across all IoU thresholds
mAP50 = ap[:, 0].mean()
mAP50_95 = ap.mean()
```

If you need more specific guidance on processing your txt files to get the inputs for this function, let me know the exact format of your annotation and detection files.
